File: multinomial/log_multinomial.rb
require_relative 'log_factorial'

def log_multinomial(x,p)
  n=0
  a=0.0
  x.zip(p).each do |xi,pi|
    n += xi
    a += xi*Math.log(pi) - log_factorial(xi)
  end
  a += log_factorial(n)
  return a
end

File: multinomial/logmultinomial.cpp
#include "logmultinomial.h"

LogMultinomial::LogBinomail() : a(0.0), N(0.0) {}
void LogMultinomial::reset() { a=0; N=0; }
void LogMultinomial::add(int x, double p) {
  N += x;
  a += x*log(p)-log_factorial(x);
}

double LogMultinomial::result() const {
  if (N > 0) {
    a += log_factorial(N);
    N = 0;
  }
  return a;
}
